File: hw2/Polynomial.cpp
#include	<iostream>
#include	<fstream>
#include	<cctype>	//using::isdigit;
#include	<cstdlib>	//atof
#include	<bits/stdc++.h>
#include	<iomanip>

using namespace std;

//to get the highest power of the input polynomial function and change the type to int
int		highestPower(float coff[10]){
	return int(coff[0]);
}

//to display the f(x) on the screen
void	display(int highestPow, float coff[10]){
	cout << "f(x) = ";
	for(int i = 0; i <= highestPow; i++){
		if(i == 0)
			cout << coff[i+1];	//i+1 because of the coff[0] is number of highest power
		else{
			if(coff[i+1] != 0){
				cout << " + " << coff[i+1];
				for(int x = i; x > 0; x--)
					cout << "*x";
			}
		}
	}
	cout << endl;
}

//compute the derivative coefficient
void	derivative(int highestPow, float coff[10], float dev_coff[]){
	if(highestPow == 0)
		dev_coff[0] = highestPow;
	else
		dev_coff[0] = highestPow-1;

	for(int i = 2; i <= highestPow+1; i++)
			dev_coff[i-1] = coff[i]*(i-1);
	
	cout << "f'(x) = ";
	for(int i = 0; i <= dev_coff[0]; i++){
			if(i == 0)
				if(dev_coff[i+1] != 0 || dev_coff[i] == 0)
					cout << dev_coff[i+1];
			else{
				if(dev_coff[i+1] != 0){
					cout << " + " << dev_coff[i+1];
					for(int x = i; x > 0; x--)
						cout << "*x";
				}
			}
	}
	cout << endl;
}

//to compute the value of f(x) and f'(x)
float	compute(int highestPow, float coff[10], float x){
	float	sum = coff[1], tmp = 0;
	for(int i = 2; i <= highestPow+1; i++){
		tmp = coff[i];
		for(int j = i-1; j > 0; j--)
				tmp = tmp * x;
		sum += tmp;
	}
	return sum;
}

int		main(){
	float	coff[10] = {0}, dev_coff[10] = {0}, poly_out = 0, poly_dev_out = 0, x = 0;
	char	buf[100] = {0}, word[100] = {0}, input[5] = {0};
	int		buf_len = 0, word_len = 0, num = 0, highestPow = 0, flag = 0, debug = 0;

	ifstream 	fin;

	while(1){
		//determine which mode is chose by user
		cout << "Two mode: if you want to input polynomial data from keyboard, please enter '1'. In contrast, if you want to input polynomial data from file, please enter '0'." << endl;
		cin >> input;

		if(atoi(input) == 0 || atoi(input) == 1){
			debug = atoi(input);
			break;
		}
		else
			cout << "Input is not legal." << endl;
	}

	if(debug == 1)
		cin.getline(buf, sizeof(buf));
	else if(debug == 0)
		fin.open("polydata.txt");

	while(fin.getline(buf, sizeof(buf)) || debug == 1){
		//keyboard mode: need to enter polynomial every time
		if(debug == 1){
			cout << "Please enter the polynomial data." << endl;
			cin.getline(buf, sizeof(buf));
		}

		cout << "Please enter the 'x' of the polynomial:" << endl;
		cin >> x;
		cin.get();	//in order to let cin.getline to get the correct input

		//initialization 
		flag = 0;	//determine if the input data is legal or not
		for(int i = 0; i < num; i++){
				coff[i] = 0;
				dev_coff[i] = 0;
		}
		num = 0;

		buf_len = strlen(buf);
		buf[buf_len] = '\0';
	
		//judge of the input polynomial is legal or not. If it is legal, change the type and put into coff[], or print the error message on the screen
		for(int i = 0; i <= buf_len ; i++){
			if(isdigit(buf[i]) || buf[i] == ' ' || buf[i] == '-' || buf[i] == '.' || buf[i] == '\n' || buf[i] == '\0'){
				if(buf[i] == '.' || isdigit(buf[i]) || buf[i] == '-'){
					if(buf[i] == '-' && word_len != 0){
						cout << "Input is not legal." << endl;
						for(int j = 0; j < word_len; j++)
								word[j] = '\0';
						word_len = 0;
						flag = 1;
						break;
					}
					word[word_len++] = buf[i];
				}
				else if(buf[i] == ' '  || buf[i] == '\n' || buf[i] == '\0'){
					coff[num++] = atof(word);
					for(int j = 0; j < word_len; j++)
							word[j] = '\0';
					word_len = 0;
				}
			}
			else{
				cout << "Input is not legal." << endl;
				for(int j = 0; j < word_len; j++)
						word[j] = '\0';
				word_len = 0;
				flag = 1;
				break;
			}
		}
		//If the amount of input is not legal, print the error message on the screen.
		if(num != (coff[0] + 2) && flag != 1){
			cout << "Input is not legal." << endl;
			flag = 1;
		}

		//subroutine
		if(flag != 1){
			highestPow = highestPower(coff);
			display(highestPow, coff);
			derivative(highestPow, coff, dev_coff);
			poly_out = compute(highestPow, coff, x);
			cout << "f(" << x << ") = " << fixed << setprecision(3) << poly_out << endl;
			poly_dev_out = compute(dev_coff[0], dev_coff, x);
			cout.unsetf(ios::fixed);
			cout << "f'(" << x << ") = " << fixed << setprecision(3) << poly_dev_out << endl;
			cout.unsetf(ios::fixed);
		}

		//initialization
		for(int i = 0; i <= buf_len; i++)
			buf[i] = '\0';
	}
	fin.close();
	cout << "Complete all calculations." << endl;
	return 0;
}

File: hw2/Polynomial.py
def is_float(value):
    try:
        float(value)
        return True
    except:
        return False

def highestPower(coff):
    return int(coff[0])

def display(highestPow, coff):
    print("f(x) = ")
    for i in highestPow:
        if i == 0:
            print(coff[i+1])
        else:
            if coff[i+1] != 0:
                print(" + ", coff[i+1])
                for x in range(i, 1, -1):
                    print("*x")
    print("\n")

def derivative(highestPow, coff, dev_coff):
    if highestPow == 0:
        dev_coff[0] = highestPow
    else:
        dev_coff[0] = highestPow - 1

    for i in range(2, highestPow+1):
        dev_coff[i-1] = coff[i]*(i-1)

    print("f'(x) = ")
    for i in dev_coff[0]:
        if i == 0:
            print(dev_coff[i+1])
        else:
            if dev_coff[i+1] != 0:
                print(" + ", dev_coff[i+1]) 
                for x in range(i, 1, -1):
                    print("*x")

    print("\n")

def compute(highestPow, coff, x):
    total = coff[1]
    tmp = 0
    for i in range(2, highestPow+1):
        tmp = coff[i]
        for j in range(i-1, 1, -1):
            tmp = tmp * x
        total += tmp
    return total


while 1:
    debug = input("The mode: if you want to input polynomial data from keyboard, please enter '1'. In contrast, if you want to input polynomial data from file, please enter '0'.\n")
    #debug = int(debug)

    if debug.isdigit() and (debug == '0' or debug == '1'):
        debug = int(debug)
        break
    else:
        print("Input is not legal.")


if debug == 1:
    buf = input("Please enter the polynomial data.")
    #buf = list(buf)
    #inp_list = buf.spilt(" ")
    de_next = 0
elif debug == 0:
    f = open('polydata.txt', 'r')
    #buf = f.readline()

while True:
    if debug == 1 and de_next == 1:
        buf = input("Please enter the polynomial data.")
        #buf = list(buf)
    elif debug == 1 and de_next == 0:
        de_next = 1
    elif debug == 0:
        buf = f.readline()
        if len(buf) == 0:
            break

    x = input("Please enter the 'x' of the polynomial.")
    x = float(x)

    flag = 0
    buf_len = len(buf)
    word_len = 0
    word = ['\0']*100
    coff = [0]*100
    dev_coff = [0]*100
    num = 0

    inp_list = buf.spilt(" ")
    inp_len = len(inp_list)

    for cal in range(0, inp_len-1):
        if is_float(inp_list[cal]) == False:
            print("Input is not legal.")
            flag = 1
            break
        else:
            coff[num] = float(inp_list[cal])
            num += 1

"""
    for i in range(0, buf_len):
        if buf[i].isdigit or buf[i] == ' ' or buf[i] == '-' or buf[i] == '.' or buf[i] == '\n' or buf[i] == '\0':
            print("test: ",buf[i], "\n")
            if buf[i] == '.' or buf[i].isdigit or buf[i] == '-':
                if buf[i] == '-' and word_len != 0:
                    #print("Input is not legal.11111")
                    for j in range(0, word_len):
                        word[j] = '\0'
                    word_len = 0
                    break;
                word[word_len] = buf[i];
                #print("word_len:", word_len)
                word_len += 1
            elif buf[i] == ' ' or buf[i] == '\n' or buf[i] == '\0':
                coff[num] = float(word)
                #print("num1:", num)
                num += 1
                for j in range(0, word_len):
                    word[j] = '\0'
                word_len = 0
        else:
            print("Input is not legal.222222")
            for j in range(0, word_len):
                word[j] = '\0'
            word_len = 0
            flag = 1
            break
"""
    print("num:",num,"coff:",coff[0],"flag:",flag,"\n")
    if num !=(coff[0] + 2) and flag != 1:
        print("Input is not legal.33333")
        flag = 1

    if flag != 1:
        highestPow = highestPower(coff)
        display(highestPow, coff)
        derivative(highestPow, coff, dev_coff)
        poly_out = compute(highestPow, coff, x)
        print("f(", x, ") = ", poly_out, "\n")
        poly_dev_out = compute(dev_coff[0], dev_coff, x)
        print("f'(", x, ") = ", poly_dev_out, "\n")

    #for i in range(0, buf_len):
       # buf =
    for i in range(0, num):
        coff[i] = 0
        dev_coff[i] = 0
    #num = 0

f.close()
print("Complete all calculation.\n")

